File: string-calculator/strign-calculator-java/src/main/java/pl/frognar/_03/StringCalculator.java
package pl.frognar._03;

import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class StringCalculator {
    public int add(String numbers) {
        if (numbers == null || numbers.isEmpty()) {
            return 0;
        }
        numbers = replaceCustomDelimitersWithComma(numbers);
        var numberList = splitAndConvertNumbers(numbers);
        AssertDoesNotContainsNegativeValues(numberList);
        return calculateSumOf(numberList);
    }

    private static final Pattern customDelimiterPattern = Pattern.compile("//(\\[?.+]?)+\n");
    private static final String escapedRegex = "\\Q%s\\E";
    private String replaceCustomDelimitersWithComma(String numbers) {
        Matcher customDelimiterMatcher = customDelimiterPattern.matcher(numbers);
        if (customDelimiterMatcher.find()) {
            var allCustomDelimiters = getAllCustomDelimiters(customDelimiterMatcher.group(1));
            for (var delimiter: allCustomDelimiters) {
                numbers = numbers.replaceAll(escapedRegex.formatted(delimiter), ",");
            }
            return customDelimiterPattern.split(numbers)[1];
        }
        return numbers;
    }

    private static List<String> getAllCustomDelimiters(String delimitersGroup) {
        delimitersGroup = delimitersGroup.replaceAll(escapedRegex.formatted("]["), " ");
        delimitersGroup = delimitersGroup.replaceAll(escapedRegex.formatted("["), "");
        delimitersGroup = delimitersGroup.replaceAll(escapedRegex.formatted("]"), "");
        return  Arrays.stream(delimitersGroup.split(" "))
                .sorted(Comparator.comparingInt(String::length).reversed())
                .toList();
    }

    private static final String defaultDelimiters = "[,\n]";
    private List<Integer> splitAndConvertNumbers(String numbers) {
        return Arrays.stream(numbers.split(defaultDelimiters)).map(Integer::parseInt).collect(Collectors.toList());
    }

    private void AssertDoesNotContainsNegativeValues(List<Integer> numbers) {
        var negatives = numbers.stream().filter(n -> n < 0).toList();
        if (!negatives.isEmpty()) {
            String negativesString = negatives
                    .stream()
                    .map(String::valueOf)
                    .collect(Collectors.joining(","));
            throw new NegativeNumbersNotAllowedException(
                    "negatives not allowed: %s".formatted(negativesString));
        }
    }

    private int calculateSumOf(List<Integer> numbers) {
        return numbers.stream().reduce(0, Integer::sum);
    }
}

File: string-calculator/string-calculator-py/07/string_calculator.py
from re import split, search
from typing import Optional


class StringCalculator:
    @classmethod
    def add(cls, numbers: Optional[str]) -> int:
        if not numbers:
            return 0
        numbers = cls.replace_custom_delimiters_with_comma(numbers)
        return sum(cls.split_and_convert(numbers))

    @staticmethod
    def replace_custom_delimiters_with_comma(numbers: str) -> str:
        custom_delimiter_search = search('//(.)\n|//\\[(.+)]\n', numbers)
        if custom_delimiter_search:
            delimiter = custom_delimiter_search.group(1)
            if not delimiter:
                delimiter = custom_delimiter_search.group(2)
            numbers = numbers[custom_delimiter_search.end():]
            numbers = numbers.replace(delimiter, ',')
        return numbers

    @staticmethod
    def split_and_convert(numbers: str):
        return [int(n) for n in split('[,\n]', numbers)]
